Cabal/Distribution/C2Hs.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+-- Based on https://github.com/gtk2hs/gtk2hs/blob/master/tools/src/Gtk2HsSetup.hs#L414
+module Distribution.C2Hs ( reorderC2Hs ) where
+
+import Prelude()
+import Distribution.Compat.Prelude
+import qualified Data.Map as M
+import qualified Data.Set as S
+import Distribution.C2Hs.Lexer
+import Distribution.ModuleName (ModuleName, components)
+import Distribution.Parsec (simpleParsec)
+import Distribution.Simple.Utils (dieNoVerbosity, findFileWithExtension, withUTF8FileContents)
+import System.FilePath (joinPath)
+
+reorderC2Hs :: [FilePath] -> [ModuleName] -> IO [ModuleName]
+reorderC2Hs dirs preMods = do
+
+  let findModule = findFileWithExtension [".chs"] dirs . joinPath . components
+
+  mFiles <- traverse findModule preMods
+
+  let preDeps = zip (fmap (ModDep []) preMods) mFiles
+
+  modDeps <- traverse extractDeps preDeps
+
+  let mods = reverse (sortTopological modDeps)
+
+  pure (moduleOriginal <$> mods)
+
+data ModDep = ModDep { moduleRequires :: [ModuleName]
+                     , moduleOriginal :: ModuleName
+                     }
+
+extractDeps :: (ModDep, Maybe FilePath) -> IO ModDep
+extractDeps (md, Nothing) = pure md
+extractDeps (md, (Just f)) = withUTF8FileContents f $ \con -> do
+  mods <- case getImports con of
+        Right ms -> case traverse simpleParsec ms of
+            Just ms' -> pure ms'
+            Nothing -> dieNoVerbosity ("Cannot parse module name in c2hs file " ++ f)
+        Left err -> dieNoVerbosity ("Cannot parse c2hs import in " ++ f ++ ": " ++ err)
+  pure (md { moduleRequires = mods })
+
+sortTopological :: [ModDep] -> [ModDep]
```

Comment:
Is there a way to `Distribution.Compat.Graph` to not reinvent the wheel?

Reply:
Yes; I've fixed this.